### PR TITLE
Tests reproing 503s (was 404s) during routerule apply - with t.Skip() until we have fix

### PR DESCRIPTION
**Pull request by ldemailly**
Test for issue #1038, errors during routingrule application

```release-note
Test for issue #1038, errors during routingrule application
```

### CODE REVIEW

**Comment by ldemailly** [comment]
got errors with:
```
bazel build //tests/e2e/tests/simple:go_default_test

./bazel-bin/tests/e2e/tests/simple/go_default_test -alsologtostderr -test.v -v 2  --skip_cleanup --namespace=e2e-2 -istioctl ~/tmp/istio-0.2.7/bin/istioctl --auth_enable
```

```
# during istio install:
customresourcedefinition "routerules.config.istio.io" created
# but then later:
 the server could not find the requested resource (post routerules.config.istio.io)
, err: exit status 1
# and pilot in crash loop with
W1006 01:30:58.604835       1 client.go:194] name conflict: ListKindConflict
I1006 01:30:58.604867       1 client.go:198] missing status condition for "routerules.config.istio.io"
```

**Comment by ldemailly** [comment]
This fails, finding 404s and 503s, the 404s need fixing probably [using this](https://envoyproxy.github.io/envoy/configuration/http_conn_man/route_config/route_config.html#config-http-conn-man-route-table) and the 503s probably some other check for "no match" ?

**Comment by ldemailly** [comment]
somehow on prow it only got 503s and no 404, probably need to make changes more than once to trigger it reliably, on my personal cluster it happens more often

/retest

**Comment by ldemailly** [comment]
to repro/see:
```
git checkout issue_1038_aka_404s
bazel build //tests/e2e/tests/simple:go_default_test
# first time:
./bazel-bin/tests/e2e/tests/simple/go_default_test -alsologtostderr -test.v -v 2  --skip_cleanup --namespace=e2e -istioctl  ~/tmp/istio-0.2.7/bin/istioctl --auth_enable --test.run  Test404sDuringChanges # will create the e2e namespace and install istio init, may get unrelated failue
# Subsequent:
./bazel-bin/tests/e2e/tests/simple/go_default_test -alsologtostderr -test.v -v 2  --skip_cleanup --skip_setup --namespace=e2e -istioctl  ~/tmp/istio-0.2.7/bin/istioctl --auth_enable --test.run  Test404sDuringChanges
# output:
14:23:24 W http.go:540> Parsed non ok code 404 (HTTP/1.1 404)
14:23:24 W http.go:540> Parsed non ok code 404 (HTTP/1.1 404)
14:23:24 W http.go:540> Parsed non ok code 404 (HTTP/1.1 404)
14:23:24 W http.go:540> Parsed non ok code 404 (HTTP/1.1 404)
14:23:24 W http.go:540> Parsed non ok code 404 (HTTP/1.1 404)
[...]
	simple1_test.go:191: Not all 146 requests were successful (map[200:136 404:10])
```

**Comment by ldemailly** [comment]
to test:
```
# first install
./bazel-bin/tests/e2e/tests/simple/go_default_test -alsologtostderr -test.v -v 2  --skip_cleanup  --namespace=e2e-isaiah -istioctl  ~/tmp/istio-0.2.7/bin/istioctl --auth_enable --test.run  Test404sDuringChanges --pilot_hub=$PILOT_HUB --pilot_tag=$PILOT_TAG
# then add --skip_setup
./bazel-bin/tests/e2e/tests/simple/go_default_test -alsologtostderr -test.v -v 2  --skip_setup --skip_cleanup  --namespace=e2e-isaiah -istioctl  ~/tmp/istio-0.2.7/bin/istioctl --auth_enable --test.run  Test404sDuringChanges --pilot_hub=$PILOT_HUB --pilot_tag=$PILOT_TAG
```

**Comment by ldemailly** [comment]
the 404s still repro locally for me but somehow they didn't in the last run since picking up new deps; they did before though, retrying:

/retest

**Comment by ldemailly** [comment]
showed up in https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/1041/e2e-suite-rbac-auth/1114/

File: tests/e2e/tests/simple/simple1_test.go [hunk 1]
```diff
@@ -148,6 +152,82 @@ func TestAuth(t *testing.T) {
...
+		URL: url,
+	}
+	res, err := fhttp.RunHTTPTest(&opts)
+	if err != nil {
+		t.Fatalf("Generating traffic via fortio failed: %v", err)
+	}
+	numRequests := res.DurationHistogram.Count
+	num200s := res.RetCodes[http.StatusOK]
+	// 1 or a handful of 503s is maybe ok, but not 1/2
+	if num200s != numRequests {
```
Comment (ZackButcher) [inline tests/e2e/tests/simple/simple1_test.go]: Based on comment, should these be compared directly or should it do something like `if num200s != (numRequests * 0.95) { ...}` or similar (i.e. check that some % of requests were successful, but not 100%).

File: tests/e2e/tests/simple/simple1_test.go [hunk 2]
```diff
@@ -148,6 +152,82 @@ func TestAuth(t *testing.T) {
...
+		},
+		URL: url,
+	}
+	res, err := fhttp.RunHTTPTest(&opts)
+	if err != nil {
+		t.Fatalf("Generating traffic via fortio failed: %v", err)
+	}
+	numRequests := res.DurationHistogram.Count
+	num200s := res.RetCodes[http.StatusOK]
+	if num200s != numRequests {
```
Comment (ZackButcher) [inline tests/e2e/tests/simple/simple1_test.go]: Same as below, should this check for slightly less than numRequests rather than numRequests exactly?

**Comment by ldemailly** [comment]
yes both fail (you can check the prow logs and see both 404s and 503s, also you can run the test individually though with new mono repo I'm not sure the instructions I put still work)

File: tests/e2e/tests/simple/simple1_test.go (hunk 1 above)
Comment (ldemailly) [reply tests/e2e/tests/simple/simple1_test.go]: it's a bit of the feature/requirement question : should changing rules for existing already deployed/up svc create any error ? in the strictest sense I think it shouldn't - there should be 0 errors, but maybe a handful is ok, my comment is somewhat outdated or reflecting what I observed (a lot of 503s)

File: tests/e2e/tests/simple/simple1_test.go (hunk 2 above)
Comment (ldemailly) [reply tests/e2e/tests/simple/simple1_test.go]: it's a bit of the feature/requirement question : should changing rules for existing already deployed/up svc create any error ? in the strictest sense I think it shouldn't - there should be 0 errors, in particular 0 404s

**Comment by ldemailly** [comment]
yes it was mostly for you to have something you can test pilot/envoy fixes against - hopefully we can indeed merge it when it starts passing

**Comment by ldemailly** [comment]
I'm not sure what "[ci skip]" is supposed to mean but we shouldn't drive anything from the subject of a PR (maybe have a label? but then it shouldn't be settable by anyone)

And in this case this PR definitely should go through all CI and should fail until we have a fix on all of them

**Comment by mandarjog** [comment]
@ldemailly I agree we should not use PR titles for this purpose.
It is some circle CI thing.

e2e tests now run on Master only, not on presubmits.
This means the test run only *after* PR is merged.
You will need to kick off e2e on this PR manually. 

@sebastienvas will this work? The expectation is that this test is run against the current Master.

**Comment by ldemailly** [comment]
we need to seriously get better at discussing changes before we make them

it's not a good idea to stop running the few e2e tests we do have!

**Comment by mandarjog** [comment]
@yutongz https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/1041/e2e-suite-rbac-auth/1734
e2e build fails because bazel is too old !

**Comment by geeknoid** [comment]
Is this PR still relevant? It dates back to October. Should it be closed?

**Comment by ldemailly** [comment]
it's still relevant because the bug is still there. once the bug is fixed it should be merged as a regression test that it is

**Comment by kyessenov** [comment]
Can this be merged and marked as non-blocking test?

**Comment by ldemailly** [comment]
I'm not sure how, it's part of simple and the rest of simple should still stay blocking

**Comment by andraxylia** [comment]
Can we have a simpler version of this test without routing rules, that only checks for 503 during normal flow?

**Comment by ldemailly** [comment]
503s without changes are a different and harder to reproduce issue. i have it reproduced in the perf setup (#3295) but you can start from this one and add another test without rule change (add, not remove/change) ?

but i suggest you just try a fix with those tests as they are
(the second one is too strict btw as mentioned above)

**Comment by andraxylia** [comment]
can you please rebase?

**Comment by ldemailly** [comment]
just did (I wonder where the github button went, in theory anyone can rebase)

**Comment by ldemailly** [comment]
Test404sDuringChanges didn't fail on circle, it may need to be rerun a few times to fail
Test503sWithPartialRules did fail as it's very strict
it got 503s for the duration of the test but maybe the health check / checking for non existent version is not a feature that works yet so I would focus on Test404sDuringChanges (which I can rename to Test503sDuringChanges)

**Comment by ldemailly** [comment]
second run on circleci failed in the 'right' test but for the wrong reason:
```
=== RUN   Test404sDuringChanges
2018-02-14T05:20:31.870466Z	info	I httprunner.go:72> Starting http test for http://10.142.1.78:31864/fortio/debug with 8 threads at 8.0 qps
2018-02-14T05:20:31.873405Z	info	W http.go:667> Parsed non ok code 404 (HTTP/1.1 404)
--- FAIL: Test404sDuringChanges (0.00s)
	simple1_test.go:188: Generating traffic via fortio failed: error 404 for http://10.142.1.78:31864/fortio/debug: "HTTP/1.1 404 Not Found\r\ndate: Wed, 14 Feb 2018 05:20:31 GMT\r\nserver: envoy\r\ncontent-length: 0\r\nx-envoy-upstream-service-time: 0\r\n\r\n"
```
and later
```
* cannot delete route-rule/istio-system/rule-3: routerules.config.istio.io "rule-3" not found
```
so it looks like the rule didn't apply at all somehow


@rkpagadala do you know why none of the prow test seem to have run e2e_simple ? (it did not too long ago)

**Comment by ldemailly** [comment]
https://circleci.com/gh/istio/istio/32508
has the right expected failure:

```
=== RUN   Test404sDuringChanges
2018-02-15T02:07:20.586620Z	info	I httprunner.go:72> Starting http test for http://10.142.1.87:32135/fortio/debug with 8 threads at 8.0 qps
Starting at 8 qps with 8 thread(s) [gomax 2] for 20s : 20 calls each (total 160)
2018-02-15T02:07:29.587076Z	info	Changing rules mid run to v1/v2
2018-02-15T02:07:29.587220Z	info	Running command /go/out/linux_amd64/release/istioctl -n istio-system create -f /go/src/istio.io/istio/tests/e2e/tests/simple/routingrule1.yaml
2018-02-15T02:07:29.684271Z	info	Command output: 
 Created config route-rule/istio-system/rule-1 at revision 989
, err: <nil>
2018-02-15T02:07:30.097539Z	info	W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
2018-02-15T02:07:30.097854Z	info	W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
2018-02-15T02:07:30.098099Z	info	W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
2018-02-15T02:07:30.098374Z	info	W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
2018-02-15T02:07:30.101881Z	info	W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
2018-02-15T02:07:30.102393Z	info	W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
2018-02-15T02:07:30.102641Z	info	W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
2018-02-15T02:07:30.102728Z	info	W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
2018-02-15T02:07:33.684424Z	info	Changing rules mid run to a/b
2018-02-15T02:07:33.684479Z	info	Running command /go/out/linux_amd64/release/istioctl -n istio-system create -f /go/src/istio.io/istio/tests/e2e/tests/simple/routingrule2.yaml
2018-02-15T02:07:33.758427Z	info	Command output: 
 Created config route-rule/istio-system/rule-2 at revision 995
, err: <nil>
2018-02-15T02:07:37.758628Z	info	Running command kubectl delete -n istio-system -f /go/src/istio.io/istio/tests/e2e/tests/simple/routingrule1.yaml
2018-02-15T02:07:37.905227Z	info	Command output: 
 routerule "rule-1" deleted
, err: <nil>
2018-02-15T02:07:37.905267Z	info	Running command kubectl delete -n istio-system -f /go/src/istio.io/istio/tests/e2e/tests/simple/routingrule2.yaml
2018-02-15T02:07:38.065967Z	info	Command output: 
 routerule "rule-2" deleted
, err: <nil>
2018-02-15T02:07:40.628022Z	info	I periodic.go:523> T005 ended after 20.009854534s : 20 calls. qps=0.9995075159600358
2018-02-15T02:07:40.628073Z	info	I periodic.go:523> T002 ended after 20.009925762s : 20 calls. qps=0.999503958079702
2018-02-15T02:07:40.628090Z	info	I periodic.go:523> T007 ended after 20.009945345s : 20 calls. qps=0.9995029799018175
2018-02-15T02:07:40.628105Z	info	I periodic.go:523> T000 ended after 20.009960399s : 20 calls. qps=0.9995022279504112
2018-02-15T02:07:40.628120Z	info	I periodic.go:523> T001 ended after 20.009975567s : 20 calls. qps=0.9995014703058183
2018-02-15T02:07:40.628135Z	info	I periodic.go:523> T004 ended after 20.009991287s : 20 calls. qps=0.9995006850899286
2018-02-15T02:07:40.628700Z	info	I periodic.go:523> T006 ended after 20.010545198s : 20 calls. qps=0.9994730179564996
2018-02-15T02:07:40.629487Z	info	I periodic.go:523> T003 ended after 20.011316214s : 20 calls. qps=0.999434509260711
Ended after 20.01139686s : 160 calls. qps=7.9954
Sleep times : count 152 avg 1.0374356 +/- 0.00512 min 1.017450595 max 1.046385411 sum 157.690218
Aggregated Function Time : count 160 avg 0.014670283 +/- 0.005075 min 0.006034764 max 0.034980807 sum 2.34724534
# range, mid point, percentile, count
>= 0.00603476 <= 0.007 , 0.00651738 , 3.12, 5
> 0.007 <= 0.008 , 0.0075 , 5.00, 3
> 0.008 <= 0.009 , 0.0085 , 6.25, 2
> 0.009 <= 0.01 , 0.0095 , 13.12, 11
> 0.01 <= 0.011 , 0.0105 , 20.62, 12
> 0.011 <= 0.012 , 0.0115 , 30.00, 15
> 0.012 <= 0.014 , 0.013 , 48.12, 29
> 0.014 <= 0.016 , 0.015 , 70.00, 35
> 0.016 <= 0.018 , 0.017 , 85.62, 25
> 0.018 <= 0.02 , 0.019 , 89.38, 6
> 0.02 <= 0.025 , 0.0225 , 95.62, 10
> 0.025 <= 0.03 , 0.0275 , 96.88, 2
> 0.03 <= 0.0349808 , 0.0324904 , 100.00, 5
# target 90% 0.0205
Code 200 : 152 (95.0 %)
Code 503 : 8 (5.0 %)
Response Header Sizes : count 160 avg 161.26875 +/- 37.12 min 0 max 174 sum 25803
Response Body/Total Sizes : count 160 avg 1540.6188 +/- 1076 min 107 max 2849 sum 246499
--- FAIL: Test404sDuringChanges (20.04s)
	simple1_test.go:193: Not all 160 requests were successful (map[200:152 503:8])
```

same in https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/1041/e2e-suite-rbac-auth/808/
```
I0215 02:16:57.137] --- FAIL: Test404sDuringChanges (21.90s)
I0215 02:16:57.137] 	simple1_test.go:193: Not all 160 requests were successful (map[503:16 200:144])
```

Happy to explain this log / what happens if anyone has a question

I filled #3509 for the prow skipping this test issue

**Comment by ldemailly** [comment]
well bad news: #3482 doesn't help:
```
--- FAIL: Test503sDuringChanges (20.30s)
	simple1_test.go:193: Not all 160 requests were successful (map[200:120 503:40])
```
/test e2e-suite-rbac-auth

File: tests/e2e/tests/simple/simple1_test.go [hunk 3]
```diff
@@ -150,6 +152,83 @@ func TestAuth(t *testing.T) {
...
+	opts.URL = url
+	res, err := fhttp.RunHTTPTest(&opts)
+	if err != nil {
+		t.Fatalf("Generating traffic via fortio failed: %v", err)
+	}
+	numRequests := res.DurationHistogram.Count
+	num200s := res.RetCodes[http.StatusOK]
+	numErrors := numRequests - num200s
+	// 1 or a handful of 503s (1 per connection) is maybe ok, but not much more
+	if numErrors > int64(opts.NumThreads)*2 {
```
Comment (ldemailly) [inline tests/e2e/tests/simple/simple1_test.go]: @ZackButcher I changed that one and open to what is the right value

**Comment by andraxylia** [comment]
@ldemailly thanks for adding this. I took the tests and used them in #4419 with alpha3 rules, please feel free to close this.

**Comment by ldemailly** [comment]
@andraxylia let me get those in shape, then maybe merge them with "t.Skip()" and you can pick them up

**Comment by ldemailly** [comment]
maybe you can make a new test with alpha3 rules but I think it's important to resolve the 503s with alpha1

**Comment by andraxylia** [comment]
We will drop alpha1 is 0.8, so I will not even try.

**Comment by ldemailly** [comment]
we are not dropping alpha1 in 0.8; 0.8 is the last version with alpha1 and alpha1 is the only stable one

**Comment by ldemailly** [comment]
btw it's pretty good right now / better - at least on circle; it only got 8 errors during rule deletion, vs a lot more errors in 0.2

/test e2e-simple

File: tests/e2e/tests/simple/a_simple_1_test.go [hunk 4]
```diff
@@ -41,6 +43,9 @@ import (
 const (
 	servicesYaml         = "tests/e2e/tests/simple/testdata/servicesToBeInjected.yaml"
 	nonInjectedYaml      = "tests/e2e/tests/simple/testdata/servicesNotInjected.yaml"
+	routingR1Yaml        = "tests/e2e/tests/simple/testdata/routingrule1.yaml"
+	routingR2Yaml        = "tests/e2e/tests/simple/testdata/routingrule2.yaml"
+	routingRNPYaml       = "tests/e2e/tests/simple/testdata/routingruleNoPods.yaml"
```
[the remark this answers is not in the document]
Comment (ldemailly) [reply tests/e2e/tests/simple/a_simple_1_test.go]: it'd be a good test of our automated conversion/translation -> v1alpha3 (https://github.com/istio/istio/issues/3621#issuecomment-383181661), but I want to get this in as is before we branch, it's been > 6 months and will allow andra and others test their fixes

**Comment by ldemailly** [comment]
All required tests passed
screen shot 2018-04-20 at 11 31 24 am

**Pull request merged**